letshelp-letsencrypt/letshelp_letsencrypt/letshelp_letsencrypt_apache.py

```diff
@@ -0,0 +1,303 @@
+
+    :param str apache_ctl: Path to `apachectl` binary
+
+
+    :returns: Path to Apache server root and main configuration file
+    :rtype: `tuple` of `str`
+
+    """
+    try:
+        proc = subprocess.Popen([apache_ctl, "-V"],
+                                stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        output, _ = proc.communicate()
+    except OSError:
+        sys.exit(_NO_APACHECTL)
+
+    server_root = config_file = ""
+    for line in output.splitlines():
+        # Relevant output lines are of the form: -D DIRECTIVE="VALUE"
+        if "HTTPD_ROOT" in line:
+            server_root = line[line.find("\"")+1:-1]
+        elif "SERVER_CONFIG_FILE" in line:
+            config_file = line[line.find("\"")+1:-1]
+
+    if not (server_root and config_file):
+        sys.exit("Unable to locate Apache configuration. Please run this "
+                 "script again and specify --server-root and --config-file")
+
+    return server_root, config_file
+
+
+def get_args():
+    """Parses command line arguments
+
+    :returns: Parsed command line options
+    :rtype: argparse.Namespace
+
+    """
+    parser = argparse.ArgumentParser(description=_DESCRIPTION)
+    parser.add_argument("-c", "--apache-ctl", default="apachectl",
+                        help="path to the `apachectl` binary")
+    parser.add_argument("-d", "--server-root",
+                        help=("location of the root directory of your Apache "
+                              "configuration"))
+    parser.add_argument("-f", "--config-file",
+                        help=("location of your main Apache configuration "
+                              "file relative to the server root"))
+    args = parser.parse_args()
+
+    # args.server_root XOR args.config_file
+    if bool(args.server_root) != bool(args.config_file):
+        sys.exit("If either --server-root and --config-file are specified, "
+                 "they both must be included")
+    elif args.server_root and args.config_file:
+        args.server_root = os.path.abspath(args.server_root)
+        args.config_file = os.path.abspath(args.config_file)
+
+        if args.config_file.startswith(args.server_root):
+            args.config_file = args.config_file[len(args.server_root)+1:]
+        else:
+            sys.exit("This script expects the Apache configuration file to be "
+                     "inside the server root")
+
+    return args
+
+
+def main():
+    """Main script execution"""
+    args = get_args()
+    if not args.server_root:
+        args.server_root, args.config_file = locate_config(args.apache_ctl)
+
+    verify_config(args)
+    tempdir = setup_tempdir(args)
+    atexit.register(lambda: shutil.rmtree(tempdir))
+    make_and_verify_selection(args.server_root, tempdir)
+
+    tarpath = os.path.join(tempdir, "config.tar.gz")
+    # contextlib.closing used for py26 support
+    with contextlib.closing(tarfile.open(tarpath, mode="w:gz")) as tar:
```

Comment:
https://docs.python.org/2/library/tarfile.html#tarfile-objects: "A TarFile object can be used as a context manager in a with statement.", so `contextlib.closing` is not necessary